src/core/inductive.ml

```diff
@@ -0,0 +1,292 @@
+                        isn't the constant TYPE. Please, raise an issue."
+               pp_term rec_typ
+      end
+  | None   ->
+      fatal pos "The type of the generated inductive principle of
+                 [%a] isn't typable. Please, raise an issue."
+        pp_term rec_typ
+
+(** [check_name_constructor i l] creates the prefixes for variables' name and
+    predicats' name according to the name of the inductive type [i] and its
+    constructors [l]. *)
+let check_name_constructor : sym -> sym list -> string * string = fun i l ->
+  (* [aux_2acc l acc1 acc2] returns the list of constructors' name which are
+     in conflict with the names of the form "x" with a number or not (stored
+     in [acc1]) or in conflict with the names of the form "p" with a number
+     or not (stored in [acc2]). *)
+  let rec aux_2acc l acc1 acc2 = match l with
+    | [] -> acc1, acc2
+    | t::q ->
+        let s = t.sym_name in
+        if Str.string_match (Str.regexp "^x[0-9]*$") s 0 then
+          aux_2acc q (s::acc1) acc2
+        else
+          if Str.string_match (Str.regexp "^p[0-9]*$") s 0 then
+            aux_2acc q acc1 (s::acc2)
+          else
+            aux_2acc q acc1 acc2
+  in
+  (* [get_prefix l e] returns a string of the form "e(i+1)" where i is the
+     maximum of the list [l]. Indeed, each element in the list [l] should
+     have the form "ei". In case of l = ["e"], the result is "e1". *)
+  let get_prefix l e = match l with
+    | [] -> e
+    | _ ->
+        let rec aux l acc = match l with
+          | [] -> acc
+          | t::q ->
+              let len = String.length t - 1 in
+              let curr_int =
+                try
+                  int_of_string (String.sub t 1 len)
+                with _ -> 0
+              in
+              if acc < curr_int then
+                aux q curr_int
+              else
+                aux q acc
+        in
+        let res = aux l (-1) in
+        e ^ (string_of_int (res+1))
+  in
+  let acc1, acc2 = aux_2acc (i::l) [] [] in
+  (get_prefix acc1 "x", get_prefix acc2 "p")
+
+(** [gen_ind_typ_codom pos ind_sym codom s] assumes that the type of [ind_sym]
+    is of the form [Π(i1:a1),...,Π(in:an), TYPE]. It then generates a [tbox]
+    similar to this type except that [TYPE] is replaced by
+    [codom [i1;...;in]]. The string [s] is the prefix of variables' name. *)
+let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> string -> tbox =
+  fun pos ind_sym codom s ->
+  let rec aux : tvar list -> term -> tbox = fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> codom (List.rev_map _Vari xs)
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b s in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
+  in aux [] !(ind_sym.sym_type)
+
+(** [fold_cons_typ pos codom get_name build_rec_hyp domrec dom ind_sym
```

Comment:
get_name does not exist anymore. Thanks to update the comment.

Reply:
The comment is not always very clear. I propose to reorganize it a little bit as follows, and do some renamings as well:
acc -> init
res -> acc
codom -> finalize
```
(** [fold_cons_typ pos codom get_name build_rec_hyp domrec dom ind_sym
cons_sym f_rec f_not_rec acc s] generates some value iteratively by going
through the structure of [cons_sym.sym_type]. In this iteration, we keep track of the variables [xs] we went through (the last variable comes first) and some accumulor [acc:'a].
There are several cases:
1) If the current type is of the form [ind_sym ts], then we call [finalize ts xs acc].
2) If the current type is a product of the form [\Pi x:ind_sym ts, b], then ...
3) If the current type is a product [\Pi x:a,b] not of the previous form, then ...
4) Any other case is an error.
```

Reply:
Done